packages/@aws-cdk/aws-servicecatalog/lib/private/association-manager.ts

```diff
@@ -0,0 +1,123 @@
+import * as cdk from '@aws-cdk/core';
+import { TagOption } from '../common';
+import { TagUpdatesOptions } from '../constraints';
+import { IPortfolio, Portfolio } from '../portfolio';
+import { IProduct, Product } from '../product';
+import { CfnPortfolioProductAssociation, CfnResourceUpdateConstraint, CfnTagOption, CfnTagOptionAssociation } from '../servicecatalog.generated';
+import { hashValues } from './util';
+import { InputValidator } from './validation';
+
+export class AssociationManager {
+  public static associateProductWithPortfolio(scope: cdk.Resource, portfolio: IPortfolio, product: IProduct) {
+    const associationKey = hashValues(portfolio.node.addr, product.node.addr, scope.stack.node.addr);
+    if (!this.associationMap.has(associationKey)) {
+      const association = new CfnPortfolioProductAssociation(scope, `PortfolioProductAssociation${associationKey}`, {
+        portfolioId: portfolio.portfolioId,
+        productId: product.productId,
+      });
+      this.associationMap.set(associationKey, new Map());
+      this.associationMap.get(associationKey).set(Constraints.ASSOCIATION, association);
+    }
+  }
+
+  public static addResourceUpdateConstraint(scope: cdk.Resource, product: IProduct, options: TagUpdatesOptions) {
+    const [pair, associationKey] = this.associationPrecheck(scope, product);
+    InputValidator.validateLength(this.generateAssocationString(scope, pair), 'description', 0, 2000, options.description);
+    if (!this.associationMap.get(associationKey).get(Constraints.RESOURCE_UPDATE)) {
+      const constraint = new CfnResourceUpdateConstraint(scope, `ResourceUpdateConstraint${associationKey}`, {
+        acceptLanguage: options.acceptLanguage,
+        description: options.description,
+        portfolioId: pair.portfolio.portfolioId,
+        productId: pair.product.productId,
+        tagUpdateOnProvisionedProduct: options.tagUpdateOnProvisionedProductAllowed || options.tagUpdateOnProvisionedProductAllowed === undefined
+          ? Allowed.ALLOWED : Allowed.NOT_ALLOWED,
+      });
+
+      constraint.addDependsOn(this.associationMap.get(associationKey).get(Constraints.ASSOCIATION));
+      this.associationMap.get(associationKey).set(Constraints.RESOURCE_UPDATE, constraint);
+    } else {
+      throw new Error(`Cannot have multiple resource update constraints for association ${this.generateAssocationString(scope, pair)}`);
+    }
+  }
+
+  public static associateTagOption(scope: cdk.Resource, resourceId: string, tagOption: TagOption) {
+    Object.keys(tagOption).forEach(key => {
+      InputValidator.validateLength(resourceId, 'TagOption key', 1, 128, key);
+      tagOption[key].forEach(value => {
+        InputValidator.validateLength(resourceId, 'TagOption value', 1, 256, value.value);
+        const tagOptionKey = hashValues(key, value.value);
+        if (!this.tagOptionMap.has(tagOptionKey)) {
+          const tO = new CfnTagOption(scope, `TagOption${tagOptionKey}`, {
```

Comment:
Weird choice for the variable name...?

```suggestion
          const cfnTagOption = new CfnTagOption(scope, `TagOption${tagOptionKey}`, {
```

Reply:
yeah, didn't even realize it was a zero and not capital 'o', fixed.